File: Common/Command.cpp
#include "Command.h"
#include "string.h"

struct CmdInfo
{
	UnityCommand id;
	const char* name;
};

static const CmdInfo infos[] = {
	{ UCOM_Shutdown, "shutdown" },
	{ UCOM_Add, "add" }, //
	{ UCOM_ChangeDescription, "changeDescription" }, //
	{ UCOM_ChangeMove, "changeMove" }, //
	{ UCOM_ChangeStatus, "changeStatus" }, //
	{ UCOM_Changes, "changes" }, //
	{ UCOM_Checkout, "checkout" }, //  
	{ UCOM_Config, "pluginConfig" }, //
	{ UCOM_DeleteChanges, "deleteChanges" }, //
	{ UCOM_Delete, "delete" }, //
	{ UCOM_Download, "download" }, //
	{ UCOM_Exit, "exit" }, //
	{ UCOM_GetLatest, "getLatest" }, //
	{ UCOM_IncomingChangeAssets, "incomingChangeAssets" }, //
	{ UCOM_Incoming, "incoming" }, //
	{ UCOM_Lock, "lock" }, //
	{ UCOM_Login, "login" }, //
	{ UCOM_Move, "move" }, //
	{ UCOM_QueryConfigParameters, "queryConfigParameters" }, //
	{ UCOM_Resolve, "resolve" }, //
	{ UCOM_RevertChanges, "revertChanges" }, //
	{ UCOM_Revert, "revert" }, //
	{ UCOM_Status, "status" }, //
	{ UCOM_Submit, "submit" }, //
	{ UCOM_Unlock, "unlock" }, //
	{ UCOM_FileMode, "filemode" }, //
	{ UCOM_CustomCommand, "customCommand" }, //
	{ UCOM_Invalid, 0 } // delimiter
};

const char* UnityCommandToString(UnityCommand c)
{
	const CmdInfo* cur = infos;
	while (cur->name && cur->id != c) { cur++; }
	return cur->name ? cur->name : "invalid";
}


UnityCommand StringToUnityCommand(const char* name)
{
	const CmdInfo* cur = infos;
	while (cur->name && strncmp(cur->name, name, 50)) { cur++; }
	return cur->id;
}

CommandException::CommandException(UnityCommand c, const std::string& about)
{ 
	m_What += UnityCommandToString(c);
	m_What += "Command";
	m_What += " => ";
	m_What += about;
}


const char* CommandException::what() const throw()
{
	return m_What.c_str();
}

File: TFSPlugin/TfsPlugin/Common/Command.cs
using System;
using System.Collections.Generic;

namespace TfsPlugin
{
    public enum UnityCommand
    {
        UCOM_Invalid,
        UCOM_Shutdown,
        UCOM_Add,
        UCOM_ChangeDescription,
        UCOM_ChangeMove,
        UCOM_ChangeStatus,
        UCOM_Changes,
        UCOM_Checkout,
        UCOM_Config,
        UCOM_DeleteChanges,
        UCOM_Delete,
        UCOM_Download,
        UCOM_Exit,
        UCOM_GetLatest,
        UCOM_IncomingChangeAssets,
        UCOM_Incoming,
        UCOM_Lock,
        UCOM_Login,
        UCOM_Move,
        UCOM_QueryConfigParameters,
        UCOM_Resolve,
        UCOM_RevertChanges,
        UCOM_Revert,
        UCOM_Status,
        UCOM_Submit,
        UCOM_Unlock,
        UCOM_FileMode,
        UCOM_CustomCommand,
    };


    public class CommandArgs : List<string>
    {

    }

    public interface ICommand<Sess, Req, Resp>
        where Req : BaseRequest
        where Resp : BaseResponse
    {
        bool Run(Sess task, Req req, Resp resp);
    }

    [Serializable]
    public class CommandException : Exception
    {
        public CommandException() { }
        public CommandException(string message) : base(message) { }
        public CommandException(string message, Exception inner) : base(message, inner) { }
        protected CommandException(
          System.Runtime.Serialization.SerializationInfo info,
          System.Runtime.Serialization.StreamingContext context)
            : base(info, context) { }
    }
}
